File: leet-code/2. Add Two Numbers/AddTwoNumbers2_2.java
public class AddTwoNumbers2_2 {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode head = new ListNode();
        ListNode cur = head;
        int carry = 0;
        while (l1 != null || l2 != null || carry != 0) {
            int sums = 0;

            if (l1 != null) {
                sums += l1.val;
                l1 = l1.next;
            }

            if (l2 != null) {
                sums += l2.val;
                l2 = l2.next;
            }

            int remainder = (carry + sums) % 10;
            carry = (carry + sums) / 10;

            cur.next = new ListNode(remainder);
            cur = cur.next;
        }

        return head.next;
    }
}

File: leet-code/2. Add Two Numbers/AddTwoNumbers2_2.kt
package com.example

class AddTwoNumbers2_2 {
    fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
        return add(l1, l2, 0)
    }

    private fun add(l1: ListNode?, l2: ListNode?, carry: Int): ListNode? {
        if (l1 == null && l2 == null) {
            return if (carry == 0) null else ListNode(carry)
        }
        val sum = (l1?.`val` ?: 0) + (l2?.`val` ?: 0)
        val node = ListNode((sum + carry) % 10)
        node.next = add(l1?.next, l2?.next, (sum + carry) / 10)
        return node
    }
}
